Move clipboard functions to a separate namespace

rachel-fenichel opened the pull request:
- Moves clipboard functions `copy`, `paste`, and `duplicate` out of `blockly.js`.
- Moves clipboard state `clipboardXml_`, `clipboardTypeCounts_`, and `clipboardSource_` out of `blockly.js`
- Aliases old functions and properties to refer to new ones.
- Updates tests that spy on specific functions.
- Updates callers inside core.

This is part of https://github.com/google/blockly/issues/5208

There's an API question here: should clipboard xml, type counts, and source be exported? The keyboard nav plugin accesses them directly, but with a note that it needs a better way to do that.

I propose that there should instead be a getter for `clipboardInfo` that returns an object containing all three of those properties, but as copies (and effectively read-only).

I also propose that we remove `clipboardXml_`, `clipboardTypeCounts_`, and `clipboardSource_` from `blockly.js` entirely and not export those three from `clipboard`.

alschmiedt reviewed (commented):
I agree with everything proposed in the PR description. We just also need to make an issue to update keyboard nav to use the new API before releasing.

rachel-fenichel commented:
Tracking -samples update in https://github.com/google/blockly-samples/issues/833

This PR also closes https://github.com/google/blockly/issues/4600

rachel-fenichel commented on core/blockly.js:
```diff
@@ -402,9 +352,7 @@ Blockly.defineBlocksWithJsonArray = function(jsonArray) {
-        Blockly.Blocks[typename] = {
-          init: Blockly.jsonInitFactory_(elem)
-        };
+        Blockly.Blocks[typename] = {init: Blockly.jsonInitFactory_(elem)};
```
Clang-format did this.

rachel-fenichel commented on core/clipboard.js:
```diff
@@ -0,0 +1,106 @@
+    typeCounts = data.typeCounts;
+  }
+};
+exports.copy = copy;
```
Done.

rachel-fenichel commented on core/blockly.js:
```diff
@@ -220,7 +206,8 @@ Blockly.deleteBlock = function(selected) {
     Blockly.hideChaff();
     if (selected.outputConnection) {
-      // Do not attempt to heal rows (https://github.com/google/blockly/issues/4832)
+      // Do not attempt to heal rows
```
clang-format did this.

rachel-fenichel commented on core/blockly.js:
```diff
@@ -136,9 +124,7 @@ Blockly.svgSize = function(svg) {
-      'Blockly.svgSize',
-      'March 2021',
-      'March 2022',
+      'Blockly.svgSize', 'March 2021', 'March 2022',
```
clang-format did this.

alschmiedt commented on core/blockly.js:
```diff
@@ -92,25 +93,12 @@ Blockly.selected = null;
- * @type {Blockly.WorkspaceSvg}
- * @private
+ * Get the current contents of the clipboard and associated metadata.
+ * @return {{xml: ?Element, source: ?Blockly.WorkspaceSvg, typeCounts: ?Object}}
```
If we could provide a better type for `typeCounts` now that this is public I think that would be really helpful.

rachel-fenichel replied:
I switched all of this to `ICopyable.CopyData`, which is an existing type.

alschmiedt commented on core/blockly.js:
```diff
@@ -92,25 +93,12 @@ Blockly.selected = null;
+ * Get the current contents of the clipboard and associated metadata.
+ * @return {{xml: ?Element, source: ?Blockly.WorkspaceSvg, typeCounts: ?Object}}
+ *     An object containing the clipboard contents and associated metadata.
+ * @public
```
I think we aren't adding `@public`.

rachel-fenichel replied:
Deleted function.

alschmiedt commented on core/blockly.js:
```diff
@@ -92,25 +93,12 @@ Blockly.selected = null;
- * @private
- */
-Blockly.clipboardTypeCounts_ = null;
+Blockly.getClipboardInfo = Blockly.clipboard.getClipboardInfo;
```
Do we need to add this here since it is a new API?

rachel-fenichel replied:
I removed this for now. However, we will have to revisit when we remove `declareLegacyNamespace` in `clipboard.js`. Since the clipboard didn't exist at all before this quarter, and I prefer to keep it internal, I may end up adding this single function at the top level.

BeksOmega commented:
Just tagging this as something I was planning on switching over to the JSO system. But since this is such a big change it may be better to wait.

rachel-fenichel commented:
@alschmiedt I made some significant changes, so please take another look.

Changes:
- Switched to storing data as an `ICopyable.CopyData` instead of storing each component individually.
- Got rid of the exported function `getClipboardInfo` and also removed it from `blockly.js`--it's not clear that it's the right API, and we have an issue tracking that.
